File: ios/Runner/RideRequestPayload.swift
import Foundation

struct RideRequestPayload {
  let rideId: String
  let pickupAddress: String
  let dropoffAddress: String
  let riderName: String?
  let title: String
  let body: String

  func toDataMap() -> [String: String] {
    var map: [String: String] = [
      "type": "new_ride_request",
      "target_user_type": "driver",
      "ride_id": rideId,
      "pickup_address": pickupAddress,
      "dropoff_address": dropoffAddress,
    ]
    if let riderName, !riderName.isEmpty {
      map["rider_name"] = riderName
    }
    return map
  }

  private static let titlePattern = try! NSRegularExpression(pattern: "^New ride request from .+$")
  private static let bodyPattern = try! NSRegularExpression(pattern: "^Pickup at .+ → .+$")
  private static let riderPattern = try! NSRegularExpression(pattern: "^New ride request from (.+)$")
  private static let routePattern = try! NSRegularExpression(pattern: "^Pickup at (.+) → (.+)$")

  static func from(userInfo: [AnyHashable: Any]) -> RideRequestPayload? {
    let data = flattenFcmData(userInfo)
    guard isForDriver(data) else { return nil }
    guard !isIncomingCall(data) else { return nil }

    let type = normalized(data["type"]) ?? normalized(data["notification_type"])
    if type == "new_ride_request" {
      return build(data: data, notificationTitle: nil, notificationBody: nil)
    }

    let rideId = normalized(data["ride_id"]) ?? normalized(data["id"]) ?? ""
    if !rideId.isEmpty {
      let status = normalized(data["status"]) ?? ""
      if status.isEmpty || status == "requested" || status == "pending" {
        return build(data: data, notificationTitle: nil, notificationBody: nil)
      }
    }

    let title = normalized(data["notification_title"])
      ?? normalized(data["title"]) ?? ""
    let body = normalized(data["notification_body"])
      ?? normalized(data["body"])
      ?? normalized(data["message"]) ?? ""

    if !title.isEmpty && looksLikeTitle(title) {
      return build(data: data, notificationTitle: title, notificationBody: body)
    }
    if !body.isEmpty && matches(bodyPattern, in: body) {
      return build(data: data, notificationTitle: title, notificationBody: body)
    }

    return nil
  }

  private static func build(
    data: [String: String],
    notificationTitle: String?,
    notificationBody: String?
  ) -> RideRequestPayload? {
    let rideId = normalized(data["ride_id"]) ?? normalized(data["id"]) ?? ""
    if rideId.isEmpty { return nil }

    var pickup = normalized(data["pickup_address"]) ?? ""
    var dropoff = normalized(data["dropoff_address"]) ?? ""
    var riderName = normalized(data["rider_name"])

    let resolvedTitle = notificationTitle?.trimmingCharacters(in: .whitespacesAndNewlines)
      ?? normalized(data["notification_title"])
      ?? normalized(data["title"]) ?? ""
    let resolvedBody = notificationBody?.trimmingCharacters(in: .whitespacesAndNewlines)
      ?? normalized(data["notification_body"])
      ?? normalized(data["body"])
      ?? normalized(data["message"]) ?? ""

    if !resolvedTitle.isEmpty,
       let match = firstMatch(riderPattern, in: resolvedTitle),
       match.count > 1 {
      let name = (resolvedTitle as NSString).substring(with: match[1]).trimmingCharacters(in: .whitespacesAndNewlines)
      if !name.isEmpty { riderName = name }
    }

    if !resolvedBody.isEmpty,
       let match = firstMatch(routePattern, in: resolvedBody),
       match.count > 2 {
      if pickup.isEmpty {
        pickup = (resolvedBody as NSString).substring(with: match[1]).trimmingCharacters(in: .whitespacesAndNewlines)
      }
      if dropoff.isEmpty {
        dropoff = (resolvedBody as NSString).substring(with: match[2]).trimmingCharacters(in: .whitespacesAndNewlines)
      }
    }

    let title: String
    if !resolvedTitle.isEmpty {
      title = resolvedTitle
    } else if let riderName, !riderName.isEmpty {
      title = "New ride request from \(riderName)"
    } else {
      title = "New ride request"
    }

    let body: String
    if !pickup.isEmpty || !dropoff.isEmpty {
      body = "Pickup at \(pickup.isEmpty ? "—" : pickup) → \(dropoff.isEmpty ? "—" : dropoff)"
    } else if !resolvedBody.isEmpty {
      body = resolvedBody
    } else {
      body = "Open the app to review this ride request."
    }

    return RideRequestPayload(
      rideId: rideId,
      pickupAddress: pickup,
      dropoffAddress: dropoff,
      riderName: riderName,
      title: title,
      body: body
    )
  }

  private static func flattenFcmData(_ userInfo: [AnyHashable: Any]) -> [String: String] {
    var data: [String: String] = [:]
    for (key, value) in userInfo {
      guard let keyString = key as? String else { continue }
      if keyString.hasPrefix("google.") || keyString.hasPrefix("gcm.") || keyString == "aps" {
        continue
      }
      if let stringValue = value as? String {
        data[keyString] = stringValue
      } else if let numberValue = value as? NSNumber {
        data[keyString] = numberValue.stringValue
      }
    }
    return data
  }

  private static func isForDriver(_ data: [String: String]) -> Bool {
    let target = normalized(data["target_user_type"]) ?? ""
    if target.isEmpty || target == "all" { return true }
    return target == "driver"
  }

  private static func isIncomingCall(_ data: [String: String]) -> Bool {
    let type = normalized(data["type"])
      ?? normalized(data["notification_type"])
      ?? normalized(data["call_type"]) ?? ""
    return type == "incoming_call"
  }

  private static func looksLikeTitle(_ title: String) -> Bool {
    let trimmed = title.trimmingCharacters(in: .whitespacesAndNewlines)
    if trimmed == "New ride request" || trimmed == "New ride request with stop(s)" {
      return true
    }
    return matches(titlePattern, in: trimmed)
  }

  private static func normalized(_ value: String?) -> String? {
    guard let value else { return nil }
    let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
    return trimmed.isEmpty ? nil : trimmed
  }

  private static func matches(_ regex: NSRegularExpression, in text: String) -> Bool {
    let range = NSRange(text.startIndex..<text.endIndex, in: text)
    return regex.firstMatch(in: text, range: range) != nil
  }

  private static func firstMatch(_ regex: NSRegularExpression, in text: String) -> [NSRange]? {
    let range = NSRange(text.startIndex..<text.endIndex, in: text)
    guard let match = regex.firstMatch(in: text, range: range) else { return nil }
    return (0..<match.numberOfRanges).map { match.range(at: $0) }
  }
}

File: android/app/src/main/kotlin/com/alphatecks/driver/messaging/RideRequestPayload.kt
package com.alphatecks.driver.messaging

import com.google.firebase.messaging.RemoteMessage

data class RideRequestPayload(
    val rideId: String,
    val pickupAddress: String,
    val dropoffAddress: String,
    val riderName: String?,
    val title: String,
    val body: String,
) {
    fun toDataMap(): Map<String, String> {
        val map = linkedMapOf(
            "type" to "new_ride_request",
            "target_user_type" to "driver",
            "ride_id" to rideId,
            "pickup_address" to pickupAddress,
            "dropoff_address" to dropoffAddress,
        )
        riderName?.takeIf { it.isNotEmpty() }?.let { map["rider_name"] = it }
        return map
    }

    companion object {
        private val titlePattern = Regex("^New ride request from .+$")
        private val bodyPattern = Regex("^Pickup at .+ → .+$")
        private val riderPattern = Regex("^New ride request from (.+)$")
        private val routePattern = Regex("^Pickup at (.+) → (.+)$")

        fun from(remoteMessage: RemoteMessage): RideRequestPayload? {
            if (!isForDriver(remoteMessage.data)) return null
            if (isIncomingCall(remoteMessage.data)) return null

            val data = remoteMessage.data
            val notification = remoteMessage.notification

            val type = data["type"]?.lowercase()?.trim()
                ?: data["notification_type"]?.lowercase()?.trim()
            if (type == "new_ride_request") {
                return build(data, notification?.title, notification?.body)
            }

            val rideId = data["ride_id"]?.trim().orEmpty().ifEmpty { data["id"]?.trim().orEmpty() }
            if (rideId.isNotEmpty()) {
                val status = data["status"]?.lowercase()?.trim().orEmpty()
                if (status.isEmpty() || status == "requested" || status == "pending") {
                    return build(data, notification?.title, notification?.body)
                }
            }

            val title = notification?.title?.trim().orEmpty()
                .ifEmpty { data["title"]?.trim().orEmpty() }
                .ifEmpty { data["notification_title"]?.trim().orEmpty() }
            val body = notification?.body?.trim().orEmpty()
                .ifEmpty { data["body"]?.trim().orEmpty() }
                .ifEmpty { data["message"]?.trim().orEmpty() }
                .ifEmpty { data["notification_body"]?.trim().orEmpty() }

            if (title.isNotEmpty() && looksLikeTitle(title)) {
                return build(data, title, body)
            }
            if (body.isNotEmpty() && bodyPattern.matches(body.trim())) {
                return build(data, title, body)
            }

            return null
        }

        fun isRideRequest(remoteMessage: RemoteMessage): Boolean = from(remoteMessage) != null

        private fun build(
            data: Map<String, String>,
            notificationTitle: String?,
            notificationBody: String?,
        ): RideRequestPayload? {
            val rideId = data["ride_id"]?.trim().orEmpty().ifEmpty { data["id"]?.trim().orEmpty() }
            if (rideId.isEmpty()) return null

            var pickup = data["pickup_address"]?.trim().orEmpty()
            var dropoff = data["dropoff_address"]?.trim().orEmpty()
            var riderName = data["rider_name"]?.trim()

            val resolvedTitle = notificationTitle?.trim().orEmpty()
                .ifEmpty { data["notification_title"]?.trim().orEmpty() }
            val resolvedBody = notificationBody?.trim().orEmpty()
                .ifEmpty { data["notification_body"]?.trim().orEmpty() }
                .ifEmpty { data["body"]?.trim().orEmpty() }
                .ifEmpty { data["message"]?.trim().orEmpty() }

            if (resolvedTitle.isNotEmpty()) {
                riderPattern.find(resolvedTitle)?.groupValues?.getOrNull(1)?.trim()?.let {
                    if (!it.isEmpty()) riderName = it
                }
            }

            if (resolvedBody.isNotEmpty()) {
                routePattern.find(resolvedBody)?.let { match ->
                    if (pickup.isEmpty()) pickup = match.groupValues[1].trim()
                    if (dropoff.isEmpty()) dropoff = match.groupValues[2].trim()
                }
            }

            val title = when {
                resolvedTitle.isNotEmpty() -> resolvedTitle
                !riderName.isNullOrEmpty() -> "New ride request from $riderName"
                else -> "New ride request"
            }
            val body = when {
                pickup.isNotEmpty() || dropoff.isNotEmpty() ->
                    "Pickup at ${pickup.ifEmpty { "—" }} → ${dropoff.ifEmpty { "—" }}"
                resolvedBody.isNotEmpty() -> resolvedBody
                else -> "Open the app to review this ride request."
            }

            return RideRequestPayload(
                rideId = rideId,
                pickupAddress = pickup,
                dropoffAddress = dropoff,
                riderName = riderName,
                title = title,
                body = body,
            )
        }

        private fun isForDriver(data: Map<String, String>): Boolean {
            val target = data["target_user_type"]?.lowercase()?.trim().orEmpty()
            if (target.isEmpty() || target == "all") return true
            return target == "driver"
        }

        private fun isIncomingCall(data: Map<String, String>): Boolean {
            val type = data["type"] ?: data["notification_type"] ?: data["call_type"]
            return type?.lowercase()?.trim() == "incoming_call"
        }

        private fun looksLikeTitle(title: String): Boolean {
            val trimmed = title.trim()
            return trimmed == "New ride request" ||
                trimmed == "New ride request with stop(s)" ||
                titlePattern.matches(trimmed)
        }
    }
}
